main.cpp

```diff
@@ -227,163 +147,92 @@ struct Numeric
                 std::cerr << "can't divide integers by zero!" << std::endl;
                 return *this;
             }
         } 
-        else if ( t < std::numeric_limits<Type>::epsilon() )
+        else if ( o < std::numeric_limits<OtherType>::epsilon() )
         {
             // if template type is less than epsilon warn about doing the division
             std::cerr << "warning: floating point division by zero!" << std::endl;
         }
 
-        *value /= t;
-        return *this;
-    }
-
-    Numeric& pow(const Type& t)
-    {
-        return powInternal(static_cast<Type>(t));
-    }
-
-    Numeric& apply( std::function<Numeric&(std::unique_ptr<Type>&)> stdFunction)
-    {
-        if(stdFunction)
-        {
-            return stdFunction(value);
-        }
-
-        return *this; 
-    }
-
-    // #5
-    Numeric& apply( void(*freeFunction)(std::unique_ptr<Type>&) )
-    {
-        if(freeFunction)
-        {
-            freeFunction(value);
-        }
-
-        return *this; 
-    }
-
-private:
-    std::unique_ptr<Type> value;  // #2
-    Numeric& powInternal(Type t)
-    {
-        *value = static_cast<Type>( std::pow(*value,t) );
-        return *this;
-    }
-};
-
-// #6
-template<>
-struct Numeric<double>
-{    
-    using Type = double;    // #3
-
-    Numeric(Type t) : value( std::make_unique<Type>(t) ) { }
-
-    ~Numeric() 
-    {
-        value = nullptr;
-    }
-
-    operator Type() const { return *value; }
-
-    Numeric& operator+=(const Type& t)
-    {
-        *value += t;
-        return *this;
-    }
-
-    Numeric& operator-=(const Type& t)
-    {
-        *value -= t;
-        return *this;
-    }
-
-    Numeric& operator*=(const Type& t)
-    {
-        *value *= t;
+        *value /= static_cast<NumericType>(o);
         return *this;
     }
 
-    Numeric& operator/=(const Type& t)
+    template<typename OtherType>
+    Numeric& pow(const OtherType& o) // #5
     {
-        if(t==0.0)
-        {
-            std::cout << "warning: floating point division by zero!\n";
-        }
-
-        *value /= t;
+        *value = static_cast<Type>( std::pow(*value,static_cast<NumericType>(o)) );
         return *this;
     }
 
-    Numeric& pow(const Type& t)
-    {
-        return powInternal(static_cast<Type>(t));
-    }
-
-    // #7
     template<typename Callable>
-    Numeric& apply( Callable callable)
+    Numeric& apply( Callable callable) //  #4
     {
         callable(value);
         return *this; 
     }
 
 private:
-    std::unique_ptr<Type> value = nullptr;  // #2
-    Numeric& powInternal(Type t)
-    {
-        *value = static_cast<Type>( std::pow(*value,t) );
-        return *this;
-    }
+    std::unique_ptr<Type> value;
 };
 
 template<typename Type> 
-void myNumericFreeFunct( std::unique_ptr<Type>& value)
+void cube( std::unique_ptr<Type>& value) // #7
 {
-    *value += 7;
+    *value = *value * *value * *value;
```

Comment:
use a local reference to make this look nicer:
```
auto& v = *value;
v = v * v * v;
```
- [ ] revise

Reply:
Done.